context.go

```diff
@@ -0,0 +1,86 @@
+package clock
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+)
+
+func ContextWithTimeout(parent context.Context, clock Clock, timeout time.Duration) (context.Context, context.CancelFunc) {
```

Comment:
Could we have ContextWithTimeoutCause and ContextWithDeadlineCause, so we can have a cause, that might help with debugging when we track context cancellations.

Reply:
CC @aaronbee because I'd like to enforce this, but maybe we still want to give people options?